From 39a35ceec30c1f233b3fc72bd6e6b8368c8fb141 Mon Sep 17 00:00:00 2001
From: Umesh Agashe <uagashe@cloudera.com>
Date: Wed, 8 Mar 2017 13:25:28 -0800
Subject: [PATCH 331/362] HBASE-17761: Test
 TestRemoveRegionMetrics.testMoveRegion fails
 intermittently because of race condition

Ref: CDH-53383
Signed-off-by: Esteban Gutierrez <esteban@apache.org>
(cherry picked from commit 3d9520b1403346cbe7a6322cf6c2632197d79f7a)

Change-Id: I05513481b1c0ca3bec9ebc7b5bb4776b91afdbe8
---
 .../apache/hadoop/hbase/HBaseTestingUtility.java   |   24 ++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java
index e46fa37..a82a622 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java
@@ -3444,6 +3444,30 @@ public class HBaseTestingUtility extends HBaseCommonTestingUtility {
   }
 
   /**
+   * Move region to destination server and wait till region is completely moved and online
+   *
+   * @param destRegion region to move
+   * @param destServer destination server of the region
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  public void moveRegionAndWait(HRegionInfo destRegion, ServerName destServer)
+      throws InterruptedException, IOException {
+    HMaster master = getMiniHBaseCluster().getMaster();
+    getHBaseAdmin().move(destRegion.getEncodedNameAsBytes(),
+        Bytes.toBytes(destServer.getServerName()));
+    while (true) {
+      ServerName serverName = master.getAssignmentManager().getRegionStates()
+          .getRegionServerOfRegion(destRegion);
+      if (serverName != null && serverName.equals(destServer)) {
+        assertRegionOnServer(destRegion, serverName, 200);
+        break;
+      }
+      Thread.sleep(10);
+    }
+  }
+
+  /**
    * Wait until all regions for a table in hbase:meta have a non-empty
    * info:server, up to 60 seconds. This means all regions have been deployed,
    * master has been informed and updated hbase:meta with the regions deployed
-- 
1.7.9.5

